packages/widgets/src/blocks/legacy-widget/edit/preview.js

```diff
@@ -13,36 +13,46 @@ import { Placeholder, Spinner, Disabled } from '@wordpress/components';
 import { __ } from '@wordpress/i18n';
 
 export default function Preview( { idBase, instance, isVisible } ) {
-	const [ iframeHeight, setIframeHeight ] = useState();
+	const [ isLoaded, setIsLoaded ] = useState( false );
 
 	// Resize the iframe on either the load event, or when the iframe becomes visible.
-	const ref = useRefEffect( ( iframe ) => {
-		function onChange() {
-			const boundingRect = iframe?.contentDocument?.body?.getBoundingClientRect();
-			if ( boundingRect ) {
-				// Include `top` in the height calculation to avoid the bottom
-				// of widget previews being cut-off. Most widgets have a
-				// heading at the top that has top margin, and the `height`
-				// alone doesn't take that margin into account.
-				setIframeHeight( boundingRect.top + boundingRect.height );
-			}
-		}
+	const ref = useRefEffect(
+		( iframe ) => {
+			// Only set height if the iframe is loaded,
+			// or it will grow to an unexpected large height in Safari if it's hidden initially.
+			if ( isLoaded ) {
+				function setHeight() {
+					iframe.style.height = `${ iframe.contentDocument.body.offsetHeight }px`;
```

Comment:
I think it'd be good to still consider something like:
```
Math.max( iframe.contentDocument.body.offsetHeight, iframe.contentDocument.documentElement.offsetHeight )
```

to get the right value. The body seems to have the right value for negative margins (I guess you could say they're collapsing inwards?), but the document for positive margins (collapsing outwards).

The rounding issues are a bit of a frustration. Could consider something like adding 1px to the value to cover that. Or switch back to `getBoundingClientRect` again.